core/domain/classifier_services.py

```diff
@@ -347,6 +349,122 @@ def mark_training_job_complete(job_id):
                                            feconf.TRAINING_JOB_STATUS_COMPLETE)
 
 
+def mark_training_job_failed(job_id):
+    """Updates the training job's status to failed.
+
+    Args:
+        job_id: str. ID of the ClassifierTrainingJob.
+    """
+    _update_classifier_training_job_status(job_id,
+                                           feconf.TRAINING_JOB_STATUS_FAILED)
+
+
+def mark_training_job_pending(job_id):
+    """Updates the training job's status to pending.
+
+    Args:
+        job_id: str. ID of the ClassifierTrainingJob.
+    """
+    _update_classifier_training_job_status(job_id,
+                                           feconf.TRAINING_JOB_STATUS_PENDING)
+
+
+def update_training_job_training_data(job_id, training_data):
+    """Checks for the existence of the model and then updates it.
+
+    Args:
+        job_id: str. ID of the ClassifierTrainingJob domain object.
+        training_data: list. The training data of the job.
+
+    Raises:
+        Exception. The ClassifierTrainingJobModel corresponding to the job_id
+            of the ClassifierTrainingJob does not exist.
+    """
+    classifier_training_job_model = (
+        classifier_models.ClassifierTrainingJobModel.get(job_id, strict=False))
+    if not classifier_training_job_model:
+        raise Exception(
+            'The ClassifierTrainingJobModel corresponding to the job_id of the'
+            'ClassifierTrainingJob does not exist.')
+
+    classifier_training_job = get_classifier_training_job_by_id(job_id)
+    classifier_training_job.update_training_data(training_data)
+    classifier_training_job.validate()
+
+    classifier_training_job_model.training_data = training_data
+    classifier_training_job_model.put()
+
+
+def update_failed_jobs(training_jobs):
+    """Updates status of job_models to failed.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        state_name: str. Name of the state.
+    """
+    for training_job in training_jobs:
+        mark_training_job_failed(training_job.job_id)
+
+
+def fetch_training_data(exp_id, state_name):
```

Comment:
I don't understand why we would do that, since the training data already exists at creation time, and there is an exploration version associated with the job. I can imagine discarding all earlier exp versions at job fetch time but it seems confusing to me to associate current training data with an old exp version.

Also, bear in mind that explorations can be reverted to earlier versions. So I think the training data and the resulting classifier ought to be always associated with a specific exploration version and that we should be really explicit about this, and my question still stands.